File: Arquitectura/P1/Coversion de Unidades_P1/ec.edu.monster.ws.cliente.web/src/main/java/ec/edu/monster/ws/cliente/web/servicios/SoapClienteServicio.java
package ec.edu.monster.ws.cliente.web.servicios;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

/**
 * Servicio que consume el web service SOAP de conversión de unidades.
 * Construye las peticiones XML manualmente (sin WSDL generado).
 * @author BRYAN-PC
 */
@Service
public class SoapClienteServicio {

    private static final String NAMESPACE = "http://ws.monster.edu.ec/";

    @Value("${soap.servicio.url}")
    private String urlServicio;

    /**
     * Autentica al usuario contra el servicio SOAP.
     */
    public String autenticar(String usuario, String clave) throws Exception {
        String soapXml = construirEnvelope(
                "<ws:autenticar>"
                + "<usuario>" + usuario + "</usuario>"
                + "<clave>" + clave + "</clave>"
                + "</ws:autenticar>");

        String respuesta = enviarPeticion(soapXml);
        return extraerValor(respuesta, "return");
    }

    /**
     * Convierte longitud.
     */
    public double convertirLongitud(double valor, String desde, String hasta) throws Exception {
        return convertir("convertirLongitud", valor, desde, hasta);
    }

    /**
     * Convierte masa.
     */
    public double convertirMasa(double valor, String desde, String hasta) throws Exception {
        return convertir("convertirMasa", valor, desde, hasta);
    }

    /**
     * Convierte temperatura.
     */
    public double convertirTemperatura(double valor, String desde, String hasta) throws Exception {
        return convertir("convertirTemperatura", valor, desde, hasta);
    }

    /**
     * Obtiene unidades disponibles.
     */
    public String obtenerUnidades(String categoria) throws Exception {
        String soapXml = construirEnvelope(
                "<ws:obtenerUnidades>"
                + "<categoria>" + categoria + "</categoria>"
                + "</ws:obtenerUnidades>");

        String respuesta = enviarPeticion(soapXml);
        return extraerValor(respuesta, "return");
    }

    // ========== Métodos privados ==========

    private double convertir(String operacion, double valor, String desde, String hasta) throws Exception {
        String soapXml = construirEnvelope(
                "<ws:" + operacion + ">"
                + "<valor>" + valor + "</valor>"
                + "<desde>" + desde + "</desde>"
                + "<hasta>" + hasta + "</hasta>"
                + "</ws:" + operacion + ">");

        String respuesta = enviarPeticion(soapXml);
        String valorStr = extraerValor(respuesta, "return");
        return Double.parseDouble(valorStr);
    }

    private String construirEnvelope(String cuerpo) {
        return "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
                + "<soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\" "
                + "xmlns:ws=\"" + NAMESPACE + "\">"
                + "<soapenv:Body>"
                + cuerpo
                + "</soapenv:Body>"
                + "</soapenv:Envelope>";
    }

    private String enviarPeticion(String soapXml) throws Exception {
        URL url = new URL(urlServicio);
        HttpURLConnection conexion = (HttpURLConnection) url.openConnection();
        conexion.setRequestMethod("POST");
        conexion.setRequestProperty("Content-Type", "text/xml; charset=utf-8");
        conexion.setDoOutput(true);
        conexion.setConnectTimeout(5000);
        conexion.setReadTimeout(5000);

        try (OutputStream os = conexion.getOutputStream()) {
            os.write(soapXml.getBytes("UTF-8"));
            os.flush();
        }

        int codigo = conexion.getResponseCode();
        InputStream is = (codigo >= 200 && codigo < 300)
                ? conexion.getInputStream() : conexion.getErrorStream();

        StringBuilder respuesta = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(is, "UTF-8"))) {
            String linea;
            while ((linea = reader.readLine()) != null) {
                respuesta.append(linea);
            }
        }
        conexion.disconnect();
        return respuesta.toString();
    }

    private String extraerValor(String xml, String etiqueta) {
        String[] patrones = {"<" + etiqueta + ">", "<ns2:" + etiqueta + ">", "<return>"};
        String[] cierres = {"</" + etiqueta + ">", "</ns2:" + etiqueta + ">", "</return>"};

        for (int i = 0; i < patrones.length; i++) {
            int inicio = xml.indexOf(patrones[i]);
            if (inicio != -1) {
                inicio += patrones[i].length();
                int fin = xml.indexOf(cierres[i], inicio);
                if (fin != -1) {
                    return xml.substring(inicio, fin);
                }
            }
        }
        return "";
    }
}

File: Arquitectura/P1/Coversion de Unidades_P1/ec.edu.monster.ws.cliente.movil/app/src/main/java/ec/edu/monster/ws/cliente/movil/servicios/SoapClienteServicio.kt
package ec.edu.monster.ws.cliente.movil.servicios

import ec.edu.monster.ws.cliente.movil.utilidades.ConstantesApp
import java.io.*
import java.net.HttpURLConnection
import java.net.URL

/**
 * Servicio SOAP cliente para Android.
 * Construye y envía peticiones SOAP XML manualmente.
 * @author BRYAN-PC
 */
class SoapClienteServicio {

    private val urlServicio = ConstantesApp.URL_SERVICIO
    private val namespace = ConstantesApp.NAMESPACE

    /**
     * Autentica al usuario contra el servicio SOAP.
     */
    fun autenticar(usuario: String, clave: String): String {
        val cuerpo = """
            <ws:autenticar>
                <usuario>$usuario</usuario>
                <clave>$clave</clave>
            </ws:autenticar>
        """.trimIndent()

        val respuesta = enviarPeticion(construirEnvelope(cuerpo))
        return extraerValor(respuesta, "return")
    }

    /**
     * Convierte longitud.
     */
    fun convertirLongitud(valor: Double, desde: String, hasta: String): Double {
        return convertir("convertirLongitud", valor, desde, hasta)
    }

    /**
     * Convierte masa.
     */
    fun convertirMasa(valor: Double, desde: String, hasta: String): Double {
        return convertir("convertirMasa", valor, desde, hasta)
    }

    /**
     * Convierte temperatura.
     */
    fun convertirTemperatura(valor: Double, desde: String, hasta: String): Double {
        return convertir("convertirTemperatura", valor, desde, hasta)
    }

    /**
     * Obtiene las unidades disponibles.
     */
    fun obtenerUnidades(categoria: String): String {
        val cuerpo = """
            <ws:obtenerUnidades>
                <categoria>$categoria</categoria>
            </ws:obtenerUnidades>
        """.trimIndent()

        val respuesta = enviarPeticion(construirEnvelope(cuerpo))
        return extraerValor(respuesta, "return")
    }

    // ========== Métodos privados ==========

    private fun convertir(operacion: String, valor: Double, desde: String, hasta: String): Double {
        val cuerpo = """
            <ws:$operacion>
                <valor>$valor</valor>
                <desde>$desde</desde>
                <hasta>$hasta</hasta>
            </ws:$operacion>
        """.trimIndent()

        val respuesta = enviarPeticion(construirEnvelope(cuerpo))
        return extraerValor(respuesta, "return").toDouble()
    }

    private fun construirEnvelope(cuerpo: String): String {
        return """<?xml version="1.0" encoding="UTF-8"?>
            <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
                xmlns:ws="$namespace">
                <soapenv:Body>$cuerpo</soapenv:Body>
            </soapenv:Envelope>""".trimIndent()
    }

    private fun enviarPeticion(soapXml: String): String {
        val url = URL(urlServicio)
        val conexion = url.openConnection() as HttpURLConnection
        conexion.requestMethod = "POST"
        conexion.setRequestProperty("Content-Type", "text/xml; charset=utf-8")
        conexion.doOutput = true
        conexion.connectTimeout = 5000
        conexion.readTimeout = 5000

        conexion.outputStream.use { os ->
            os.write(soapXml.toByteArray(Charsets.UTF_8))
            os.flush()
        }

        val codigo = conexion.responseCode
        val inputStream = if (codigo in 200..299) conexion.inputStream else conexion.errorStream

        val respuesta = inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
        conexion.disconnect()
        return respuesta
    }

    private fun extraerValor(xml: String, etiqueta: String): String {
        val patrones = arrayOf("<$etiqueta>", "<ns2:$etiqueta>", "<return>")
        val cierres = arrayOf("</$etiqueta>", "</ns2:$etiqueta>", "</return>")

        for (i in patrones.indices) {
            val inicio = xml.indexOf(patrones[i])
            if (inicio != -1) {
                val inicioValor = inicio + patrones[i].length
                val fin = xml.indexOf(cierres[i], inicioValor)
                if (fin != -1) return xml.substring(inicioValor, fin)
            }
        }
        return ""
    }
}
